File: sem4/jpp/list4/cpp/Philosopher.cpp
#include "Philosopher.hpp"

#include <iostream>
#include <thread>

Philosopher::Philosopher(const int id, std::mutex &leftChopstick, std::mutex &rightChopstick, const int mealLimit, std::mutex &printingMutex)
    : id_(id)
    , leftChopstick_(leftChopstick)
    , rightChopstick_(rightChopstick)
    , mealLimit_(mealLimit)
    , printingMutex_(printingMutex)
{
    std::srand(std::time(nullptr));
}

void Philosopher::operator()() const {
    for (int i = 0; i < mealLimit_; i++) {
        think();
        pickUpChopstick(leftChopstick_);
        pickUpChopstick(rightChopstick_);
        eat();
        putDownChopstick(leftChopstick_);
        putDownChopstick(rightChopstick_);
    }
}

void Philosopher::think() const {
    print("Philosopher " + std::to_string(id_) + " is thinking.");
    sleepRandomDuration();
    print("Philosopher " + std::to_string(id_) + " stopped thinking.");
}

void Philosopher::eat() const {
    print("Philosopher " + std::to_string(id_) + " is eating.");
    sleepRandomDuration();
    print("Philosopher " + std::to_string(id_) + " stopped eating.");
}

void Philosopher::pickUpChopstick(std::mutex &chopstick) const {
    sleepRandomDuration();
    chopstick.lock();
}

void Philosopher::putDownChopstick(std::mutex &chopstick) const {
    sleepRandomDuration();
    chopstick.unlock();
}

void Philosopher::print(const std::string &text) const {
    printingMutex_.lock();
    std::cout << text << '\n';
    printingMutex_.unlock();
}

void Philosopher::sleepRandomDuration() const {
    const auto randomDuration = std::chrono::milliseconds(
        rand() % lowerDurationBoundMilliseconds_ + upperDurationBoundMilliseconds_ + 1
    );
    std::this_thread::sleep_for(randomDuration);
}

File: sem4/jpp/list4/go/philosopher/philosopher.go
package philosopher

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

const (
	lowerDurationBound int           = 50
	upperDurationBound int           = 100
	durationUnit       time.Duration = time.Millisecond
)

var rng = rand.New(rand.NewSource(time.Now().UnixNano()))

type Philosopher struct {
	id             int
	leftChopstick  *sync.Mutex
	rightChopstick *sync.Mutex
	mealsLimit     int
	waitGroup      *sync.WaitGroup
}

func NewPhilosopher(id int, leftChopstick *sync.Mutex, rightChopstick *sync.Mutex, mealsLimit int, waitGroup *sync.WaitGroup) *Philosopher {
	return &Philosopher{id, leftChopstick, rightChopstick, mealsLimit, waitGroup}
}

func (p Philosopher) Dine() {
	for i := 0; i < p.mealsLimit; i++ {
		p.think()
		p.pickUpChopstick(p.leftChopstick)
		p.pickUpChopstick(p.rightChopstick)
		p.eat()
		p.putDownChopstick(p.leftChopstick)
		p.putDownChopstick(p.rightChopstick)
	}
	p.waitGroup.Done()
}

func (p Philosopher) think() {
	fmt.Printf("Philosopher %v is thinking.\n", p.id)
	p.sleepRandomDuration()
	fmt.Printf("Philosopher %v stopped thinking.\n", p.id)
}

func (p Philosopher) eat() {
	fmt.Printf("Philosopher %v is eating.\n", p.id)
	p.sleepRandomDuration()
	fmt.Printf("Philosopher %v stopped eating.\n", p.id)
}

func (p Philosopher) pickUpChopstick(chopstick *sync.Mutex) {
	p.sleepRandomDuration()
	chopstick.Lock()
}

func (p Philosopher) putDownChopstick(chopstick *sync.Mutex) {
	p.sleepRandomDuration()
	chopstick.Unlock()
}

func (p Philosopher) sleepRandomDuration() {
	var randomDuration = time.Duration(rng.Intn(lowerDurationBound) + upperDurationBound)
	time.Sleep(randomDuration * durationUnit)
}
